model/map/src/main/java/org/keycloak/models/map/storage/QueryParameters.java

```diff
@@ -69,6 +70,21 @@ public QueryParameters<M> orderBy(SearchableModelField<M> searchableModelField,
         return this;
     }
 
+    @Override
+    public boolean equals(Object o) {
+        if (this == o) return true;
+        if (o == null || getClass() != o.getClass()) return false;
+        QueryParameters<?> that = (QueryParameters<?>) o;
+        // there is currently no equals method for the ModelCriteriaBuilder, take its String representation as a substitute.
+        return Objects.equals(offset, that.offset) && Objects.equals(limit, that.limit) && Objects.equals(orderBy, that.orderBy) && Objects.equals(mcb.toString(), that.mcb.toString());
+    }
+
+    @Override
+    public int hashCode() {
+        // there is currently no equals method for the ModelCriteriaBuilder, take its String representation as a substitute.
```

Comment:
There is no guarantee that `toString` would be defined on the respective `ModelCriteriaBuilder`.
Please update JPA `ModelCriteriaBuilder` to include proper equals / hashcode instead.

Reply:
Hi @hmlnarik - I pushed the next commit that doesn't rely on that comparison any more. Instead it first maps it to a JPA criteria query and then extracts the caching key from there. 

You mention that `ModelCriteriaBuilder` doesn't need to define a `toString`. At the same time `QueryParameters` always uses a `DefaultModelCriteria` which does have a `toString()`. Should the `QueryParameters` constructor be changed to accept a different class?

Reply:
Thank you for the changes, please consider this comment resolved from my POV.

> Should the QueryParameters constructor be changed to accept a different class?

Not now, but it is possible in the future.

If `JpaModelCriteriaBuilder` defined `equals` and `hashCode`, perhaps we could leverage those instead of converting criteria to a `String`. On the other hand I see it as a potential next-step optimization, and would like to see the code in the current shape to be merged, as I believe it should already have visible impact.

Reply:
I gave it a thought to make `JpaModelCriteriaBuilder` implement `equals` and `hashCode`, but then abandoned it when I didn't find a simple way to make `JpaPredicateFunction` (one of the children `JpaModelCriteriaBuilder`) implement them for all of its variants. The corresponding element in `DefaultModelCriteria`, in the `ModelCriteriaNode`, implements `toString` which I used in the previous attempt.

Reply:
@hmlnarik - with #15038, the impact of the caching is now smaller and stated at be start of this issue; yet there is still some impact on every call on the admin UI.

Reply:
I believe there is also impact of this PR on areas not explicitly affected by #15038 like all kinds of sessions which I suspect significant enough for incorporating this PR. Eventually would be nice to see this suspicion confirmed by a benchmark run.